documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,183 @@
+
+```c
+// "EVENTPIPE_V1"
+char[15] handshake = { 0x00, 0x00, 0x00, 0x45, 0x56, 0x45, 0x4e, 0x54, 0x50, 0x49, 0x50, 0x45, 0x5f, 0x56, 0x31 };
+```
+
+The `Accept` function in the runtime will immediately try to read 15 bytes of data and then use that determine what protocol we are speaking.  Once it has been determined, the runtime will respond with that handshake.
+
+#### Errors
+* runtime: 
+  * read failure -> end connection; no response
+  * unknown protocol -> ERROR response // TODO: flesh out errors
+* Client:
+  * no response -> simply have a timeout
+  * unkown protocol error -> fix your client
+
+### Headers
+
+#### Requirements:
+* small - shouldn't affect write times or read times
+* simple - flat and blittable
+* descriptive - self describing
+* extensible - ability for us to add funcitonality without necessarily incrementing protocol version
+
+Every packet will start with a header and every header will start with a size.  This makes it easy to read the data directly out of the stream no matter the size or contents of the packet.
+
+> Taking inspiration from Mono's Soft-Debugger protocol, I am splitting command codes into command sets and commands, allowing command codes to be reused between command sets, e.g., 0x01,0x01 and 0x02,0x01 would be different commands despite sharing the same explicit value for the command code.
+
+```c
+// size = 4 + 1 + 1 + 1 = 7 bytes
+struct header
+{
+    uint16 size; // size of packet = size of header + payload
+    char   flag; // command or response
+    char   command_set;
+    char   command;
+}
+```
+
+### Packets
+
+#### Requirements:
+* non-breaking - Wrapping current implementation in these packets shouldn't add more work than simply pulling the payload out, i.e., packet is FULLY opaque and orthogonal to the header.
+
+Packets will contain a header and 0 or one payload.  The payload will completely opaque with the client and runtime agreeing on some format/encoding of the data.  The size will be known as it must be `packet.header.size - sizeof(header)`.
+
+<table>
+  <tr>
+    <th>1</th>
+    <th>2</th>
+    <th>3</th>
+    <th>4</th>
+    <th>5</th>
+    <th>6</th>
+    <th>7</th>
+    <th>8</th>
+    <th>9</th>
+    <th>10</th>
+    <th>11</th>
+    <th>...</th>
+    <th>size - 1 </th>
+    <th>size</th>
+  </tr>
+  <tr>
+    <td colspan="7">header</td>
+    <td colspan="7">payload</td>
+  </tr>
+  <tr>
+    <td colspan="4">size</td>
+    <td>flag</td>
+    <td>command_set</td>
+    <td>command</td>
+    <td colspan="7">payload</td>
+  </tr>
+</table>
+
+The simplest of command/reply packets will only contain a header and therefore be only 7 bytes long.
+
+### Commands Sets
+
+#### Requirements:
+// TODO
+
+```c
+enum command_sets : char
+{
+    EVENT_PIPE = 0x01,
+    // future
+}
+```
+
+### `EVENT_PIPE` Commands
+
+#### Requirements:
+* minimal
+* as close to a no-op with current code as possible
+
+
+```c
+enum event_pipe_commands : char
+{
```

Comment:
GET_STREAM_SESSION - I'm not sure what scenario this enables?
PAUSE_STREAM_SESSION - depending on what you had in mind, this can probably be emulated by either:
   a) stop reading from the stream and let the events buffer in the runtime
   b) use UPDATE to turn off all the providers, later use update again to turn them back on

Reply:
I'm imagining that an EventPipe session is where these commands will be sent and received, and that the client will _not_ be creating a new session to communicate new commands.  These commands just represent the base interactions I would expect to be able to perform against a session:
* start
* stop
* close
* update config
* get current config

(now that I type that out, since I made start, stop, _and_ close commands, pause seems redundant and unnecessary)

Reply:
I'm still not sure why we need get current config? If you created the session wouldn't you know how you configured it?

Reply:
I approached this the way that RESTful services work when it comes to resources: I should be able to create, read, update, and delete my perf stream config.  For simplicity I suppose it is not necessary, but ultimately it _should_ be cheap to implement and may be of use when we light up multi-session.  Suppose there are 5 clients, and client 6 comes along and wants to check what other clients are tracing.

Reply:
That sounds like a reasonable thing to do in the future, but I wouldn't preemptively work on it now. Preview6 is packed with work to do and none of our tools need this capability in order to meet their goals.